src/go/k8s/pkg/resources/statefulset_scale.go

```diff
@@ -0,0 +1,392 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	redpandav1alpha1 "github.com/redpanda-data/redpanda/src/go/k8s/apis/redpanda/v1alpha1"
+	adminutils "github.com/redpanda-data/redpanda/src/go/k8s/pkg/admin"
+	"github.com/redpanda-data/redpanda/src/go/k8s/pkg/labels"
+	"github.com/redpanda-data/redpanda/src/go/rpk/pkg/api/admin"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	DecommissionRequeueDuration = time.Second * 10
+)
+
+// handleScaling is called to detect cases of replicas change and apply them to the cluster
+func (r *StatefulSetResource) handleScaling(ctx context.Context) error {
+	if r.pandaCluster.Status.Decommission != nil {
+		if r.pandaCluster.Status.Decommission.NodeID == nil {
+			// Special type of decommission for nodes that probably never started
+			return r.handleDirectDownscale(ctx)
+		}
+
+		decommissionTargetReplicas := r.pandaCluster.Status.Decommission.Ordinal
+		if r.pandaCluster.Spec.GetReplicas() > decommissionTargetReplicas {
+			// Decommissioning can also be canceled and we need to recommission
+			return r.handleRecommission(ctx)
+		}
+		return r.handleDecommission(ctx)
+	}
+
+	if r.pandaCluster.Status.Replicas == 0 {
+		// Initialize the status replicas
+		r.pandaCluster.Status.Replicas = r.pandaCluster.Spec.GetReplicas()
+		return r.Status().Update(ctx, r.pandaCluster)
+	}
+
+	if r.pandaCluster.Spec.GetReplicas() == r.pandaCluster.Status.Replicas {
+		// No changes to replicas, we do nothing here
+		return nil
+	}
+
+	if r.pandaCluster.Spec.GetReplicas() > r.pandaCluster.Status.Replicas {
+		r.logger.Info("Upscaling cluster", "replicas", r.pandaCluster.Spec.GetReplicas())
+		// Upscaling request: this is already handled by Redpanda, so we just increase status replicas
+		return setTargetRedpandaReplicas(ctx, r, r.pandaCluster, r.pandaCluster.Spec.GetReplicas(), r.logger)
+	}
+
+	// Start decommissioning process
+	targetOrdinal := r.pandaCluster.Status.Replicas - 1
+	r.logger.Info("Start decommission of last broker node", "ordinal", targetOrdinal)
+	return r.markDecommissioningStatus(ctx, targetOrdinal)
+}
+
+// handleDecommission manages cases of node decommissioning
+func (r *StatefulSetResource) handleDecommission(ctx context.Context) error {
+	targetReplicas := r.pandaCluster.Status.Decommission.Ordinal
+	r.logger.Info("Handling cluster in decommissioning phase", "target replicas", targetReplicas)
+
+	fullAdminAPI, err := r.getAdminAPIClient(ctx)
```

Comment:
There are cases where I use an admin API client that can reach every node of the cluster and other cases where I target a specific node to get the node ID. I guess it will become just adminAPI once I remove the distinction between nodeId and ordinal